File: Baekjoon OJ/02110 - 공유기 설치.py
"""
    Code written on Jan 9, 2021
    출처: Baekjoon OJ, https://www.acmicpc.net
    02110번: 공유기 설치
    link: https://www.acmicpc.net/problem/2110
"""
from sys import stdin


num_houses, supply = map(int, stdin.readline().split())
houses = [int(stdin.readline()) for _ in range(num_houses)]
houses.sort()

min_distance = 1
max_distance = (houses[-1]-houses[0]) // (supply-1)

while min_distance <= max_distance:
    mid_distance = (min_distance + max_distance) // 2
    demand = 1
    base_point = houses[0]

    for house in houses[1:]:
        if house - base_point >= mid_distance:
            demand += 1
            base_point = house
        if demand == supply:
            min_distance = mid_distance + 1
            break
    else:
        max_distance = mid_distance - 1

print(max_distance)

File: Baekjoon OJ/02110 - 공유기 설치.java
/*
    Code written on Feb 21, 2021
    출처: Baekjoon OJ, https://www.acmicpc.net
    02110번: 공유기 설치
    link: https://www.acmicpc.net/problem/2110
*/
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());
        int[] houses = new int[n];

        for (int i = 0; i < n; i++)
            houses[i] = Integer.parseInt(br.readLine());
        Arrays.sort(houses);

        int left = 1;
        int right = (houses[n - 1] - houses[0]) / (c - 1);
        while (left <= right) {
            int mid = (left + right) / 2;
            int installed = 1;
            int baseHouse = houses[0];

            for (int targetHouse: houses) {
                if (targetHouse - baseHouse >= mid) {
                    baseHouse = targetHouse;
                    installed++;
                }
            }

            if (installed < c)
                right = mid - 1;
            else
                left = mid + 1;
        }

        System.out.print(right);
    }
}
